packages-exp/auth-exp/src/core/user/additional_user_info.test.ts

```diff
@@ -0,0 +1,215 @@
+          rawUserInfo: rawUserInfoWithLogin
+        });
+        expect(fromIdTokenResponse(idResponse)!.profile).to.eql(
+          userProfileWithLogin
+        );
+      });
+
+      it('for missing JSON', () => {
+        const idResponse = idTokenResponse({ providerId: ProviderId.FACEBOOK });
+        expect(fromIdTokenResponse(idResponse)!.profile).to.be.empty;
+      });
+    });
+
+    describe('determines new-user status', () => {
+      it('for new users by token response', () => {
+        const idResponse = idTokenResponse({
+          providerId: ProviderId.FACEBOOK,
+          isNewUser: true
+        });
+        expect(fromIdTokenResponse(idResponse)!.isNewUser).to.be.true;
+      });
+
+      it('for new users by toolkit response kind', () => {
+        const idResponse = idTokenResponse({
+          providerId: ProviderId.FACEBOOK,
+          kind: IdTokenResponseKind.SignupNewUser
+        });
+        expect(fromIdTokenResponse(idResponse)!.isNewUser).to.be.true;
+      });
+
+      it('for old users', () => {
+        const idResponse = idTokenResponse({ providerId: ProviderId.FACEBOOK });
+        expect(fromIdTokenResponse(idResponse)!.isNewUser).to.be.false;
+      });
+    });
+
+    describe('creates generic AdditionalUserInfo', () => {
+      it('for custom auth', () => {
+        const idResponse = idTokenResponse({
+          providerId: ProviderId.CUSTOM,
+          rawUserInfo: rawUserInfoWithLogin
+        });
+        const {
+          isNewUser,
+          providerId,
+          username,
+          profile
+        } = fromIdTokenResponse(idResponse)!;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.be.null;
+        expect(username).to.be.null;
+        expect(profile).to.eq(profile);
+      });
+
+      it('for anonymous auth', () => {
+        const idResponse = idTokenResponse({
+          providerId: ProviderId.ANONYMOUS,
+          rawUserInfo: rawUserInfoWithLogin
+        });
+        const {
+          isNewUser,
+          providerId,
+          username,
+          profile
+        } = fromIdTokenResponse(idResponse)!;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.be.null;
+        expect(username).to.be.null;
+        expect(profile).to.eq(profile);
+      });
+
+      it('for missing provider IDs in response but not in token', () => {
+        const {
+          isNewUser,
+          providerId,
+          username,
+          profile
+        } = fromIdTokenResponse(
+          idTokenResponse({ rawUserInfo: rawUserInfoWithLogin })
+        )!;
+        expect(isNewUser).to.be.false;
+        expect(providerId).to.eq(ProviderId.FACEBOOK);
+        expect(username).to.be.null;
+        expect(profile).to.eq(profile);
+      });
+    });
+
+    describe('returns null', () => {
+      it('for missing provider IDs', () => {
+        const idResponse = idTokenResponse({});
+        expect(fromIdTokenResponse(idResponse)!).to.be.null;
```

Comment:
how does this test work? you're asserting that it's non-null and then checking that it is null?

Reply:
heh yeah the `!` "assertion" doesn't actually do anything but mollify the TS compiler. In any case @scottcrossen you can remove it in this case

Reply:
Oops yeah my regex find-replace acted out